File: frontend/js/company.js
document.addEventListener("DOMContentLoaded", () => {
  // Tab switching
  document.querySelectorAll(".tab-btn").forEach((btn) => {
    btn.addEventListener("click", () => {
      document.querySelectorAll(".tab-btn").forEach((b) => b.classList.remove("active"));
      document.querySelectorAll(".tab-content").forEach((c) => c.classList.remove("active"));
      btn.classList.add("active");
      document.getElementById(`${btn.dataset.tab}-tab`).classList.add("active");
      if (btn.dataset.tab === "applicants") loadJobsForApplicants();
    });
  });

  loadCompanyProfile();
  loadStats();
  loadMyJobs();
  loadJobsForShortlist();

  // Company profile save
  document.getElementById("company-profile-form").addEventListener("submit", async (e) => {
    e.preventDefault();
    const data = {
      company_name: document.getElementById("profile-company-name").value,
      manager_name: document.getElementById("profile-manager-name").value,
      designation: document.getElementById("profile-designation").value,
    };
    const msg = document.getElementById("company-profile-message");
    try {
      await api.createOrUpdateCompanyProfile(data);
      msg.textContent = "Profile saved successfully!";
      msg.style.color = "green";
      setTimeout(() => { msg.textContent = ""; }, 3000);
      loadStats();
    } catch (err) {
      msg.textContent = "Error: " + err.message;
      msg.style.color = "red";
    }
  });

  // Post job
  document.getElementById("post-job-form").addEventListener("submit", async (e) => {
    e.preventDefault();
    const title = document.getElementById("job-title").value;
    const skills = document.getElementById("job-skills").value;
    const cgpa = document.getElementById("job-min-cgpa").value;
    const topN = document.getElementById("job-top-n").value;
    if (!validateNotEmpty(title)) { showFormError(e.target, "Job title is required."); return; }
    if (!validateNotEmpty(skills)) { showFormError(e.target, "At least one skill is required."); return; }
    if (!validateCGPA(cgpa)) { showFormError(e.target, "CGPA must be between 0 and 10."); return; }
    try {
      await api.addJob({
        title,
        requirements: skills.split(",").map((s) => s.trim()).join(", "),
        min_cgpa: parseFloat(cgpa),
        top_n: parseInt(topN) || 10,
      });
      alert("Job posted successfully!");
      e.target.reset();
      loadMyJobs();
      loadStats();
      loadJobsForShortlist();
      loadJobsForApplicants();
    } catch (err) {
      showFormError(e.target, err.message);
    }
  });

  // Shortlist form
  document.getElementById("shortlist-form").addEventListener("submit", async (e) => {
    e.preventDefault();
    const jobId = document.getElementById("shortlist-job-select").value;
    if (!jobId) { showFormError(e.target, "Please select a job first."); return; }
    try {
      const data = await api.companyShortlist(jobId);
      renderShortlist(data);
    } catch (err) {
      document.getElementById("shortlist-results").innerHTML = `<p class="error">Error: ${err.message}</p>`;
    }
  });

  // Load applicants button
  document.getElementById("load-applicants-btn").addEventListener("click", async () => {
    const jobId = document.getElementById("applicant-job-select").value;
    if (!jobId) { alert("Please select a job first."); return; }
    loadApplicants(jobId);
  });

  document.getElementById("refresh-jobs").addEventListener("click", loadMyJobs);
});

// ─── Stats ───────────────────────────────────────────────────────────────────
async function loadStats() {
  const el = document.getElementById("company-stats");
  if (!el) return;
  try {
    const d = await api.companyDashboard();
    const s = d.stats || {};
    el.innerHTML = `
      <div class="stat-card"><span class="stat-value">${s.my_jobs ?? 0}</span><span class="stat-label">My Jobs</span></div>
      <div class="stat-card"><span class="stat-value">${s.total_jobs_posted ?? 0}</span><span class="stat-label">Total Jobs</span></div>
    `;
  } catch (e) { el.innerHTML = ""; }
}

// ─── My Jobs ─────────────────────────────────────────────────────────────────
async function loadMyJobs() {
  const c = document.getElementById("jobs-list");
  c.innerHTML = '<p class="loading">Loading your jobs...</p>';
  try {
    const jobs = await api.getMyJobs();
    if (!jobs.length) { c.innerHTML = '<p class="info">You haven\'t posted any jobs yet.</p>'; return; }
    c.innerHTML = jobs.map((j) => `
      <div class="list-item">
        <div class="list-item-header"><strong>${j.title}</strong><span class="badge">Min CGPA: ${j.min_cgpa}</span></div>
        <div class="list-item-body">Requirements: ${j.requirements || "N/A"}</div>
        <div class="list-item-footer">
          <button class="btn-primary btn-sm" onclick="viewApplicantsForJob(${j.id}, '${j.title.replace(/'/g, "\\'")}')">View Applicants</button>
        </div>
      </div>`).join("");
  } catch (e) { c.innerHTML = `<p class="error">Error: ${e.message}</p>`; }
}

function viewApplicantsForJob(jobId, jobTitle) {
  // Switch to applicants tab and load
  document.querySelectorAll(".tab-btn").forEach((b) => b.classList.remove("active"));
  document.querySelectorAll(".tab-content").forEach((c) => c.classList.remove("active"));
  document.querySelector("[data-tab='applicants']").classList.add("active");
  document.getElementById("applicants-tab").classList.add("active");
  loadJobsForApplicants(() => {
    document.getElementById("applicant-job-select").value = jobId;
    loadApplicants(jobId);
  });
}

// ─── Applicants ───────────────────────────────────────────────────────────────
async function loadJobsForApplicants(cb) {
  const sel = document.getElementById("applicant-job-select");
  if (!sel) return;
  try {
    const jobs = await api.getMyJobs();
    sel.innerHTML = '<option value="">— Select a job to view applicants —</option>' +
      jobs.map((j) => `<option value="${j.id}">${j.title}</option>`).join("");
    if (cb) cb();
  } catch (e) { sel.innerHTML = '<option value="">Error loading jobs</option>'; }
}

const STATUS_LABELS = {
  applied: "📋 Applied",
  under_review: "🔍 Under Review",
  shortlisted: "⭐ Shortlisted",
  interview: "📅 Interview Scheduled",
  rejected: "❌ Rejected",
  offer: "🎉 Offer Sent",
};

async function loadApplicants(jobId) {
  const c = document.getElementById("applicants-list");
  c.innerHTML = '<p class="loading">Loading applicants...</p>';
  try {
    const data = await api.getJobApplicants(jobId);
    if (!data.applicants || !data.applicants.length) {
      c.innerHTML = '<p class="info">No applicants yet for this job.</p>';
      return;
    }
    c.innerHTML = `<h4 style="margin-bottom:12px;">${data.job_title} — ${data.total_applicants} Applicant(s)</h4>` +
      data.applicants.map((a) => {
        const statusClass = `status-${a.status}`;
        const statusLabel = STATUS_LABELS[a.status] || a.status;
        const interviewHtml = a.interview ? `
          <div style="background:#f3e5f5;padding:8px 12px;border-radius:8px;margin-top:8px;font-size:13px;">
            📅 <strong>Interview:</strong> ${a.interview.date} at ${a.interview.time} — ${a.interview.mode}
            ${a.interview.link ? `| <a href="${a.interview.link}" target="_blank">Join</a>` : ""}
          </div>` : "";
        const offerHtml = a.offer ? `
          <div style="background:#e8f5e9;padding:8px 12px;border-radius:8px;margin-top:8px;font-size:13px;">
            🎉 <strong>Offer Sent:</strong> ${a.offer.position} — CTC: ${a.offer.ctc}
            <span class="status-badge" style="margin-left:6px;">${a.offer.status}</span>
          </div>` : "";
        return `
          <div class="applicant-card">
            <div class="applicant-header">
              <div>
                <strong>${a.student_name}</strong>
                <span style="color:#666;font-size:13px;margin-left:8px;">${a.student_email}</span>
              </div>
              <span class="status-badge ${statusClass}">${statusLabel}</span>
            </div>
            <div style="font-size:13px;color:#444;margin-bottom:6px;">
              CGPA: <strong>${a.student_cgpa}</strong> &nbsp;|&nbsp;
              Match: <strong>${a.match_percentage}%</strong> &nbsp;|&nbsp;
              Skills: ${a.student_skills || "N/A"}
              ${a.resume_url ? ` &nbsp;|&nbsp; <a href="${api.getFileUrl(a.resume_url)}" target="_blank">📄 Resume</a>` : ""}
            </div>
            ${interviewHtml}
            ${offerHtml}
            <div class="action-btns">
              <button class="btn-xs btn-review" onclick="updateStatus(${a.application_id}, 'under_review', ${jobId})">🔍 Under Review</button>
              <button class="btn-xs btn-shortlist" onclick="updateStatus(${a.application_id}, 'shortlisted', ${jobId})">⭐ Shortlist</button>
              <button class="btn-xs btn-interview" onclick="openInterviewModal(${a.application_id})">📅 Schedule Interview</button>
              <button class="btn-xs btn-offer" onclick="openOfferModal(${a.application_id})">🎉 Send Offer</button>
              <button class="btn-xs btn-reject" onclick="updateStatus(${a.application_id}, 'rejected', ${jobId})">❌ Reject</button>
            </div>
          </div>`;
      }).join("");
  } catch (e) { c.innerHTML = `<p class="error">Error: ${e.message}</p>`; }
}

async function updateStatus(appId, status, jobId) {
  try {
    await api.updateApplicationStatus(appId, status);
    loadApplicants(jobId);
  } catch (e) { alert("Error: " + e.message); }
}

// ─── Interview Modal ──────────────────────────────────────────────────────────
function openInterviewModal(appId) {
  document.getElementById("interview-app-id").value = appId;
  // Default to tomorrow
  const tomorrow = new Date(); tomorrow.setDate(tomorrow.getDate() + 1);
  document.getElementById("interview-date").value = tomorrow.toISOString().split("T")[0];
  document.getElementById("interview-time").value = "10:00";
  document.getElementById("interview-link").value = "";
  document.getElementById("interview-notes").value = "";
  document.getElementById("interview-modal").classList.add("open");
}

async function submitInterview() {
  const appId = document.getElementById("interview-app-id").value;
  const data = {
    interview_date: document.getElementById("interview-date").value,
    interview_time: document.getElementById("interview-time").value,
    mode: document.getElementById("interview-mode").value,
    link: document.getElementById("interview-link").value || null,
    notes: document.getElementById("interview-notes").value || null,
  };
  if (!data.interview_date || !data.interview_time) { alert("Date and time are required."); return; }
  try {
    await api.scheduleInterview(appId, data);
    closeModal("interview-modal");
    alert("Interview scheduled successfully! The student has been notified.");
    const jobId = document.getElementById("applicant-job-select").value;
    if (jobId) loadApplicants(jobId);
  } catch (e) { alert("Error: " + e.message); }
}

// ─── Offer Modal ──────────────────────────────────────────────────────────────
function openOfferModal(appId) {
  document.getElementById("offer-app-id").value = appId;
  document.getElementById("offer-position").value = "";
  document.getElementById("offer-ctc").value = "";
  document.getElementById("offer-modal").classList.add("open");
}

async function submitOffer() {
  const appId = document.getElementById("offer-app-id").value;
  const position = document.getElementById("offer-position").value.trim();
  const ctc = document.getElementById("offer-ctc").value.trim();
  if (!position || !ctc) { alert("Position and CTC are required."); return; }
  try {
    await api.sendOffer(appId, { position, ctc });
    closeModal("offer-modal");
    alert("Offer sent! The student has been notified via dashboard and email.");
    const jobId = document.getElementById("applicant-job-select").value;
    if (jobId) loadApplicants(jobId);
  } catch (e) { alert("Error: " + e.message); }
}

function closeModal(id) {
  document.getElementById(id).classList.remove("open");
}
// Close modal on backdrop click
document.addEventListener("click", (e) => {
  if (e.target.classList.contains("modal-overlay")) {
    e.target.classList.remove("open");
  }
});

// ─── Shortlist Tool ───────────────────────────────────────────────────────────
async function loadJobsForShortlist() {
  const sel = document.getElementById("shortlist-job-select");
  if (!sel) return;
  try {
    const jobs = await api.getMyJobs();
    sel.innerHTML = '<option value="">Select one of your jobs</option>' +
      jobs.map((j) => `<option value="${j.id}">${j.title}</option>`).join("");
  } catch (e) { sel.innerHTML = '<option value="">Error loading</option>'; }
}

function renderShortlist(data) {
  const c = document.getElementById("shortlist-results");
  if (!data.results || !data.results.length) { c.innerHTML = '<p class="info">No matching candidates.</p>'; return; }
  c.innerHTML = `<h4>Shortlist for: ${data.job}</h4>` + data.results.map((r, i) => `
    <div class="list-item">
      <div class="list-item-header"><strong>#${i + 1}. ${r.name}</strong><span class="badge badge-success">Score: ${Number(r.score || 0).toFixed(2)}</span></div>
      <div class="list-item-body">
        CGPA: ${r.cgpa} | Skills: ${r.skills || "N/A"}
        ${r.resume_url ? ` | <a href="${api.getFileUrl(r.resume_url)}" target="_blank" rel="noopener noreferrer">View Resume</a>` : ""}
      </div>
    </div>`).join("");
}

// ─── Company Profile ──────────────────────────────────────────────────────────
async function loadCompanyProfile() {
  try {
    const profile = await api.getCompanyProfile();
    if (profile) {
      document.getElementById("profile-company-name").value = profile.company_name || "";
      document.getElementById("profile-manager-name").value = profile.manager_name || "";
      document.getElementById("profile-designation").value = profile.designation || "";
    }
  } catch (err) { console.log("No profile found or error loading profile."); }
}

File: backend/app/routes/company.py
from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional

from ..deps import company_required, get_db
from ..models import Job, Student, StudentUser, ProfileView, Notification, CompanyProfile, JobApplication, Interview, Offer
from ..schemas import CompanyProfileCreate, CompanyProfileResponse
from ..matching import score_student_for_job
from ..utils.email import send_email

router = APIRouter(prefix="/company", tags=["Company"])


# ---------- Pydantic request bodies ----------

class ApplicationStatusUpdate(BaseModel):
    status: str  # under_review | shortlisted | rejected


class InterviewSchedule(BaseModel):
    interview_date: str   # "YYYY-MM-DD"
    interview_time: str   # "HH:MM"
    mode: str             # "Google Meet" | "In-Person" | "Zoom" etc.
    link: Optional[str] = None
    notes: Optional[str] = None


class OfferCreate(BaseModel):
    position: str
    ctc: str              # "10 LPA"


# ---------- Helpers ----------

def _get_company_name(db: Session, email: str) -> str:
    profile = db.query(CompanyProfile).filter(CompanyProfile.owner_email == email).first()
    return profile.company_name if profile else email


def _notify_student(db: Session, student_email: str, message: str):
    db.add(Notification(student_email=student_email, message=message))


# ---------- Dashboard & Profile ----------

@router.get("/dashboard")
def company_dashboard(company=Depends(company_required), db: Session = Depends(get_db)):
    profile = db.query(CompanyProfile).filter(CompanyProfile.owner_email == company.email).first()
    company_name = profile.company_name if profile else "Company"

    my_jobs = db.query(Job).filter(Job.created_by == company.email).count()
    total_jobs = db.query(Job).count()
    return {
        "message": "Welcome to Company Dashboard!",
        "user_email": company.email,
        "stats": {
            "my_jobs": my_jobs,
            "total_jobs_posted": total_jobs,
        },
    }


@router.get("/my-jobs")
def get_my_jobs(company=Depends(company_required), db: Session = Depends(get_db)):
    return db.query(Job).filter(Job.created_by == company.email).all()


@router.get("/profile", response_model=CompanyProfileResponse)
def get_company_profile(company=Depends(company_required), db: Session = Depends(get_db)):
    profile = db.query(CompanyProfile).filter(CompanyProfile.owner_email == company.email).first()
    if not profile:
        raise HTTPException(status_code=404, detail="Profile not found")
    return profile


@router.post("/profile", response_model=CompanyProfileResponse)
def create_or_update_company_profile(
    profile_data: CompanyProfileCreate,
    company=Depends(company_required),
    db: Session = Depends(get_db)
):
    profile = db.query(CompanyProfile).filter(CompanyProfile.owner_email == company.email).first()
    if profile:
        profile.company_name = profile_data.company_name
        profile.manager_name = profile_data.manager_name
        profile.designation = profile_data.designation
    else:
        profile = CompanyProfile(
            owner_email=company.email,
            company_name=profile_data.company_name,
            manager_name=profile_data.manager_name,
            designation=profile_data.designation
        )
        db.add(profile)
    db.commit()
    db.refresh(profile)
    return profile


# ---------- Applicant Management ----------

@router.get("/applicants/{job_id}")
def get_job_applicants(
    job_id: int,
    company=Depends(company_required),
    db: Session = Depends(get_db)
):
    """List all applicants for a job posted by this company."""
    job = db.query(Job).filter(Job.id == job_id, Job.created_by == company.email).first()
    if not job:
        raise HTTPException(status_code=404, detail="Job not found or not owned by you")

    applications = db.query(JobApplication).filter(JobApplication.job_id == job_id).all()

    result = []
    for app in applications:
        student = db.query(Student).filter(Student.owner_email == app.student_email).first()
        if not student:
            continue

        # Fetch interview info if any
        interview = db.query(Interview).filter(Interview.application_id == app.id).first()
        interview_info = None
        if interview:
            interview_info = {
                "date": interview.interview_date,
                "time": interview.interview_time,
                "mode": interview.mode,
                "link": interview.link,
                "notes": interview.notes,
            }

        # Fetch offer info if any
        offer = db.query(Offer).filter(Offer.application_id == app.id).first()
        offer_info = None
        if offer:
            offer_info = {
                "id": offer.id,
                "position": offer.position,
                "ctc": offer.ctc,
                "status": offer.status,
            }

        result.append({
            "application_id": app.id,
            "student_email": app.student_email,
            "student_name": student.name,
            "student_skills": student.skills,
            "student_cgpa": student.cgpa,
            "resume_url": student.resume_url,
            "match_percentage": app.match_percentage,
            "status": app.status,
            "applied_at": app.applied_at.isoformat(),
            "interview": interview_info,
            "offer": offer_info,
        })

    result.sort(key=lambda x: x["match_percentage"], reverse=True)

    return {
        "job_id": job.id,
        "job_title": job.title,
        "total_applicants": len(result),
        "applicants": result,
    }


@router.post("/application/{app_id}/status")
def update_application_status(
    app_id: int,
    body: ApplicationStatusUpdate,
    background_tasks: BackgroundTasks,
    company=Depends(company_required),
    db: Session = Depends(get_db)
):
    """Update application status: under_review | shortlisted | rejected."""
    allowed = {"under_review", "shortlisted", "rejected"}
    if body.status not in allowed:
        raise HTTPException(status_code=400, detail=f"Status must be one of: {allowed}")

    app = db.query(JobApplication).filter(JobApplication.id == app_id).first()
    if not app:
        raise HTTPException(status_code=404, detail="Application not found")

    job = db.query(Job).filter(Job.id == app.job_id, Job.created_by == company.email).first()
    if not job:
        raise HTTPException(status_code=403, detail="You do not own this job")

    app.status = body.status
    company_name = _get_company_name(db, company.email)

    status_labels = {
        "under_review": "is now Under Review",
        "shortlisted": "has been Shortlisted ⭐",
        "rejected": "has been Rejected",
    }
    message = f"Your application for {job.title} at {company_name} {status_labels[body.status]}"
    _notify_student(db, app.student_email, message)

    if body.status == "shortlisted":
        student = db.query(Student).filter(Student.owner_email == app.student_email).first()
        if student:
            html = f"""
            <h2>🎉 You've been Shortlisted!</h2>
            <p>Dear {student.name},</p>
            <p>Congratulations! You have been shortlisted for <strong>{job.title}</strong> at <strong>{company_name}</strong>.</p>
            <p>Your match score: {app.match_percentage}%</p>
            <p>Please log in to the placement portal to see further updates.</p>
            """
            background_tasks.add_task(send_email, app.student_email, f"Shortlisted for {job.title}", html)

    db.commit()
    return {"message": f"Application status updated to '{body.status}'", "application_id": app_id}


@router.post("/application/{app_id}/schedule-interview")
def schedule_interview(
    app_id: int,
    body: InterviewSchedule,
    background_tasks: BackgroundTasks,
    company=Depends(company_required),
    db: Session = Depends(get_db)
):
    """Schedule an interview for an applicant."""
    app = db.query(JobApplication).filter(JobApplication.id == app_id).first()
    if not app:
        raise HTTPException(status_code=404, detail="Application not found")

    job = db.query(Job).filter(Job.id == app.job_id, Job.created_by == company.email).first()
    if not job:
        raise HTTPException(status_code=403, detail="You do not own this job")

    # Remove existing interview if rescheduling
    existing = db.query(Interview).filter(Interview.application_id == app_id).first()
    if existing:
        db.delete(existing)

    interview = Interview(
        application_id=app_id,
        job_id=app.job_id,
        student_email=app.student_email,
        company_email=company.email,
        interview_date=body.interview_date,
        interview_time=body.interview_time,
        mode=body.mode,
        link=body.link,
        notes=body.notes,
    )
    db.add(interview)
    app.status = "interview"

    company_name = _get_company_name(db, company.email)
    message = (
        f"📅 Interview Scheduled for {job.title} at {company_name} | "
        f"Date: {body.interview_date} | Time: {body.interview_time} | Mode: {body.mode}"
    )
    _notify_student(db, app.student_email, message)

    student = db.query(Student).filter(Student.owner_email == app.student_email).first()
    if student:
        link_line = f'<p><strong>Link:</strong> <a href="{body.link}">{body.link}</a></p>' if body.link else ""
        notes_line = f"<p><strong>Notes:</strong> {body.notes}</p>" if body.notes else ""
        html = f"""
        <h2>📅 Interview Scheduled</h2>
        <p>Dear {student.name},</p>
        <p>An interview has been scheduled for your application to <strong>{job.title}</strong> at <strong>{company_name}</strong>.</p>
        <table style="border-collapse:collapse;">
          <tr><td style="padding:4px 12px;"><strong>Date</strong></td><td>{body.interview_date}</td></tr>
          <tr><td style="padding:4px 12px;"><strong>Time</strong></td><td>{body.interview_time}</td></tr>
          <tr><td style="padding:4px 12px;"><strong>Mode</strong></td><td>{body.mode}</td></tr>
        </table>
        {link_line}
        {notes_line}
        <p>Best of luck!</p>
        """
        background_tasks.add_task(send_email, app.student_email, f"Interview Scheduled – {job.title}", html)

    db.commit()
    return {"message": "Interview scheduled successfully", "application_id": app_id}


@router.post("/application/{app_id}/send-offer")
def send_offer(
    app_id: int,
    body: OfferCreate,
    background_tasks: BackgroundTasks,
    company=Depends(company_required),
    db: Session = Depends(get_db)
):
    """Send a job offer to an applicant."""
    app = db.query(JobApplication).filter(JobApplication.id == app_id).first()
    if not app:
        raise HTTPException(status_code=404, detail="Application not found")

    job = db.query(Job).filter(Job.id == app.job_id, Job.created_by == company.email).first()
    if not job:
        raise HTTPException(status_code=403, detail="You do not own this job")

    # Remove existing offer if resending
    existing = db.query(Offer).filter(Offer.application_id == app_id).first()
    if existing:
        db.delete(existing)

    offer = Offer(
        application_id=app_id,
        job_id=app.job_id,
        student_email=app.student_email,
        company_email=company.email,
        position=body.position,
        ctc=body.ctc,
    )
    db.add(offer)
    app.status = "offer"

    company_name = _get_company_name(db, company.email)
    message = f"🎉 Offer Received from {company_name} for {body.position} – CTC: {body.ctc}"
    _notify_student(db, app.student_email, message)

    student = db.query(Student).filter(Student.owner_email == app.student_email).first()
    if student:
        html = f"""
        <h2>🎉 Congratulations! You have received a Job Offer</h2>
        <p>Dear {student.name},</p>
        <p><strong>{company_name}</strong> has extended an offer to you:</p>
        <table style="border-collapse:collapse;">
          <tr><td style="padding:4px 12px;"><strong>Position</strong></td><td>{body.position}</td></tr>
          <tr><td style="padding:4px 12px;"><strong>CTC</strong></td><td>{body.ctc}</td></tr>
        </table>
        <p>Please log in to the placement portal to Accept or Reject this offer.</p>
        """
        background_tasks.add_task(send_email, app.student_email, f"Job Offer – {body.position} at {company_name}", html)

    db.commit()
    return {"message": "Offer sent successfully", "application_id": app_id}


# ---------- Legacy shortlist endpoint ----------

@router.get("/shortlist/{job_id}")
def company_shortlist(
    job_id: int,
    background_tasks: BackgroundTasks,
    company=Depends(company_required),
    db: Session = Depends(get_db),
):
    job = db.query(Job).filter(Job.id == job_id, Job.created_by == company.email).first()
    if not job:
        raise HTTPException(status_code=404, detail="Job not found or not owned by you")

    students = db.query(Student).all()
    results = []
    shortlisted_students = []

    for s in students:
        scoring = score_student_for_job(s, job)
        student_data = {
            "student_id": s.id,
            "name": s.name,
            "skills": s.skills,
            "cgpa": s.cgpa,
            "resume_url": s.resume_url,
            "score": scoring if isinstance(scoring, (int, float)) else scoring.get("score", scoring),
        }
        results.append(student_data)
        if isinstance(scoring, dict) and scoring.get("overall_match_percentage", 0) >= 70:
            shortlisted_students.append(s)

    ranked = sorted(results, key=lambda x: x["score"], reverse=True)

    for student in shortlisted_students:
        if student.owner_email:
            notification = Notification(
                student_email=student.owner_email,
                message=f"You were shortlisted for {job.title}!"
            )
            db.add(notification)
            html = f"""
            <h2>Congratulations! You've been shortlisted</h2>
            <p>Dear {student.name}, you were shortlisted for <strong>{job.title}</strong>.</p>
            """
            background_tasks.add_task(send_email, student.owner_email, f"Shortlisted for {job.title}", html)

    db.commit()
    return {
        "job": job.title,
        "results": ranked,
        "shortlisted_count": len(shortlisted_students),
        "total_students": len(students)
    }


@router.post("/profile/view/{student_id}")
def log_company_profile_view(
    student_id: int,
    company=Depends(company_required),
    db: Session = Depends(get_db)
):
    student = db.query(Student).filter(Student.id == student_id).first()
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    db.add(ProfileView(student_id=student_id, viewer_email=company.email, viewer_role="company"))
    db.add(Notification(student_email=student.owner_email, message=f"Your profile was viewed by {company.email} (Company)"))
    db.commit()
    return {"message": "Profile view logged"}


@router.get("/student/{student_id}")
def get_student_profile(
    student_id: int,
    company=Depends(company_required),
    db: Session = Depends(get_db)
):
    student = db.query(Student).filter(Student.id == student_id).first()
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    db.add(ProfileView(student_id=student_id, viewer_email=company.email, viewer_role="company"))
    db.add(Notification(student_email=student.owner_email, message=f"Your profile was viewed by {company.email} (Company)"))
    db.commit()
    return {
        "id": student.id,
        "name": student.name,
        "skills": student.skills,
        "cgpa": student.cgpa,
        "resume_url": student.resume_url,
        "has_resume": bool(student.resume_url)
    }
